File: apps/ios/Loonext/Core/Model/ApiKeys.swift
import Foundation

/// #243 — workspace API keys. Twin of the Android `core/model/ApiKeys.kt`.
///
/// The token is not a property here, because it is not a field in any response
/// but the one that mints it. It exists outside the caller's own app exactly
/// once, in the 201, as `MintedApiKey.token_once`.
struct ApiKey: Codable, Identifiable, Hashable, Sendable {
    let id: String
    let name: String
    /// The first twelve characters, so three keys can be told apart.
    let token_prefix: String
    var scopes: [String] = []
    var created_at: String?
    /// The field that makes switching one off safe: is anything still calling?
    var last_used_at: String?
    var revoked_at: String?
    var expires_at: String?
}

struct ApiKeyList: Codable, Sendable {
    var keys: [ApiKey] = []
    var cap: Int = 0
    /// Live keys only — revoking makes room, so the cap counts what is on.
    var live: Int = 0
}

struct MintedApiKey: Codable, Sendable {
    let key: ApiKey
    let token_once: String
}

struct CreateApiKeyBody: Codable, Sendable {
    let name: String
    let scopes: [String]
}

File: apps/android/app/src/main/kotlin/com/loonext/android/core/model/ApiKeys.kt
package com.loonext.android.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * #243 — workspace API keys.
 *
 * The token is not a property here, because it is not a field in any response
 * but the one that mints it. It exists outside the caller's own app exactly
 * once, in the 201, as [MintedApiKey.tokenOnce].
 */
@Serializable
data class ApiKey(
    val id: String,
    val name: String,
    /** The first twelve characters, so three keys can be told apart. */
    @SerialName("token_prefix") val tokenPrefix: String,
    val scopes: List<String> = emptyList(),
    @SerialName("created_at") val createdAt: String? = null,
    /** The field that makes switching one off safe: is anything still calling? */
    @SerialName("last_used_at") val lastUsedAt: String? = null,
    @SerialName("revoked_at") val revokedAt: String? = null,
    @SerialName("expires_at") val expiresAt: String? = null,
)

@Serializable
data class ApiKeyList(
    val keys: List<ApiKey> = emptyList(),
    val cap: Int = 0,
    /** Live keys only — revoking makes room, so the cap counts what is on. */
    val live: Int = 0,
)

@Serializable
data class MintedApiKey(
    val key: ApiKey,
    @SerialName("token_once") val tokenOnce: String,
)

@Serializable
data class CreateApiKeyBody(
    val name: String,
    val scopes: List<String>,
)
